### ISSUE TITLE
Theano and other GPU app conflict

### ISSUE THREAD

**Issue by eduardohenriquearnold** [Theano/Theano#4740]
I am working on a computer vision application developed with OpenCV in C++, we use a camera that comes with a third-party library which uses CUDA for image processing. I also use Keras for Deep Learning with Theano as backend. I am calling all Keras functionality from C++ through the Python C++ API.

The problem is that whenever I am using both the Camera and the Keras procedures, I receive the following errors:
```
Using Theano backend.
Using gpu device 0: NVS 510 (CNMeM is disabled, cuDNN not available)
Traceback (most recent call last):
  File "./DeepClassifier.py", line 107, in classify
    res = model.predict(im)
  File "/usr/local/lib/python2.7/dist-packages/keras/models.py", line 464, in predict
    return self.model.predict(x, batch_size=batch_size, verbose=verbose)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/training.py", line 1155, in predict
    batch_size=batch_size, verbose=verbose)
  File "/usr/local/lib/python2.7/dist-packages/keras/engine/training.py", line 855, in _predict_loop
    batch_outs = f(ins_batch)
  File "/usr/local/lib/python2.7/dist-packages/keras/backend/theano_backend.py", line 531, in __call__
    return self.function(*inputs)
  File "/usr/local/lib/python2.7/dist-packages/theano/compile/function_module.py", line 871, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/usr/local/lib/python2.7/dist-packages/theano/gof/link.py", line 314, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/usr/local/lib/python2.7/dist-packages/theano/compile/function_module.py", line 859, in __call__
    outputs = self.fn()
RuntimeError: cublasSgemm failed (14) an internal operation failed
 unit=0 N=0, c.dims=[1 512], a.dim=[1 3600], alpha=%f, beta=%f, a=%p, b=%p, c=%p sa_0=%d, sa_1=%d, sb_0=%d, sb_1=%d, sc_0=%d, sc_1=%d
Apply node that caused the error: GpuDot22(GpuElemwise{Composite{(i0 * (i1 + Abs(i1)))}}[(0, 1)].0, dense_1_W)
Toposort index: 14
Inputs types: [CudaNdarrayType(float32, matrix), CudaNdarrayType(float32, matrix)]
Inputs shapes: [(1, 3600), (3600, 512)]
Inputs strides: [(0, 1), (512, 1)]
Inputs values: ['not shown', 'not shown']
Outputs clients: [[GpuElemwise{Composite{(i0 * ((i1 + i2) + Abs((i1 + i2))))}}[(0, 1)](CudaNdarrayConstant{[[ 0.5]]}, GpuDot22.0, GpuDimShuffle{x,0}.0)]]

Debugprint of the apply node: 
GpuDot22 [id A] <CudaNdarrayType(float32, matrix)> ''   
 |GpuElemwise{Composite{(i0 * (i1 + Abs(i1)))}}[(0, 1)] [id B] <CudaNdarrayType(float32, matrix)> ''   
 | |CudaNdarrayConstant{[[ 0.5]]} [id C] <CudaNdarrayType(float32, (True, True))>
 | |GpuReshape{2} [id D] <CudaNdarrayType(float32, matrix)> ''   
 |   |GpuFromHost [id E] <CudaNdarrayType(float32, 4D)> ''   
 |   | |flatten_input_1 [id F] <TensorType(float32, 4D)>
 |   |MakeVector{dtype='int64'} [id G] <TensorType(int64, vector)> ''   
 |     |Shape_i{0} [id H] <TensorType(int64, scalar)> ''   
 |     | |flatten_input_1 [id F] <TensorType(float32, 4D)>
 |     |Elemwise{IntDiv}[(0, 0)] [id I] <TensorType(int64, scalar)> ''   
 |       |Prod{acc_dtype=int64} [id J] <TensorType(int64, scalar)> ''   
 |       | |MakeVector{dtype='int64'} [id K] <TensorType(int64, vector)> ''   
 |       |   |Shape_i{0} [id H] <TensorType(int64, scalar)> ''   
 |       |   |Shape_i{1} [id L] <TensorType(int64, scalar)> ''   
 |       |   | |flatten_input_1 [id F] <TensorType(float32, 4D)>
 |       |   |Shape_i{2} [id M] <TensorType(int64, scalar)> ''   
 |       |   | |flatten_input_1 [id F] <TensorType(float32, 4D)>
 |       |   |Shape_i{3} [id N] <TensorType(int64, scalar)> ''   
 |       |     |flatten_input_1 [id F] <TensorType(float32, 4D)>
 |       |Shape_i{0} [id H] <TensorType(int64, scalar)> ''   
 |dense_1_W [id O] <CudaNdarrayType(float32, matrix)>

Storage map footprint:
 - dense_1_W, Shared Input, Shape: (3600, 512), ElemSize: 4 Byte(s), TotalSize: 7372800 Byte(s)
 - dense_2_W, Shared Input, Shape: (512, 256), ElemSize: 4 Byte(s), TotalSize: 524288 Byte(s)
 - GpuElemwise{Composite{(i0 * (i1 + Abs(i1)))}}[(0, 1)].0, Shape: (1, 3600), ElemSize: 4 Byte(s), TotalSize: 14400 Byte(s)
 - flatten_input_1, Input, Shape: (1, 1, 60, 60), ElemSize: 4 Byte(s), TotalSize: 14400 Byte(s)
 - dense_1_b, Shared Input, Shape: (512,), ElemSize: 4 Byte(s), TotalSize: 2048 Byte(s)
 - dense_2_b, Shared Input, Shape: (256,), ElemSize: 4 Byte(s), TotalSize: 1024 Byte(s)
 - dense_3_W, Shared Input, Shape: (256, 1), ElemSize: 4 Byte(s), TotalSize: 1024 Byte(s)
 - dense_3_b, Shared Input, Shape: (1,), ElemSize: 4 Byte(s), TotalSize: 4 Byte(s)
 - CudaNdarrayConstant{[[ 0.5]]}, Shape: (1, 1), ElemSize: 4 Byte(s), TotalSize: 4 Byte(s)
 TotalSize: 7929992 Byte(s) 0.007 GB
 TotalSize inputs: 7915592 Byte(s) 0.007 GB
```
If am using Keras alone (without initializing the camera object), no errors occur. Also, if I disable Theano GPU support, it all runs fine, but with significant performance loss.

I have also tried running two process in parallel (one which uses the camera and other that uses Keras) and both work just fine.

Given these information, my hypothesis is that when calling the Python functionality from C++ I am running everything under the same PID, so there is no distinction of CUDA contexts between Theano objects and the camera 3rd party algorithm. 

Does anyone have a suggestion on how to proceed here?

**Comment by nouiz** [comment]
You need to get information of how the 3rd party use the GPU. How it
initialize the GPU and how it create the context.

Do you want both to have the same GPU context? Otherwise, you will need to
load one, save and remove the context, then load the second software, safe
and remove the context. Then set the good GPU context before calling each
software function.

As least, this is my guess.

**Comment by eduardohenriquearnold** [comment]
Thanks for the quick reply. Unfortunately it is a 3rd party library with closed source, so I have no way of knowing how they initialize the GPU, but I have access to the following function `CUcontext getCUDAContext() `

I believe that the **cublasSgemm** error is caused because both this 3rd party library and Theano are sharing the GPU within the same context, since they run under the same process. So I would like to force them to use different contexts.

I am not sure I can modify the 3rd party CUDA context, though. Do you know a way of ensuring Theano has a different CUDA context?

**Comment by nouiz** [comment]
Did you try to init first Theano, then init the 3rd party and the reverse?
Maybe one order work.

For context, how it work is that it is a stack of context. So you can load
one software, pop the context, load the other software and set the good
context before calling it.

I didn't try it, but this is how normal multi-context work for Theano
multi-GPU in the same process work.

CUDA have function to pop and push the context on the stack.

**Comment by eduardohenriquearnold** [comment]
Thanks for the help, I solved the issue by wrapping the Theano calls through C++ in CUDA context Push/Pop, as can be seen here:

```
	cuCtxPushCurrent(theanoctx);

	PyObject *nparr = pyopencv_from(img);
	PyObject *pargs = PyTuple_New(1);
	PyTuple_SetItem(pargs, 0, nparr);
	PyObject* pres = PyObject_CallObject(classifyf, pargs);
	PyErr_Print();
	Py_DECREF(pargs);

	cuCtxPopCurrent(&theanoctx);	

	double res = PyFloat_AsDouble(pres);

```

Note that I create `theanoctx` on class construction. I hope this can be of use for other users as well. Thanks for the help.

**Comment by nouiz** [comment]
Thanks for the code, I'm pretty it can help other people.

**Comment by XapaJIaMnu** [comment]
Hey,

This is marked as fixed, but I am experiencing a problem where theano loads a C library which uses the GPU. I found out that if I ask my C library to use a different GPU from theano (the C library calls cudaSetDevice to achieve that) from the one theano is using, theano crashes. I got to this issue because my crash output is similar. Do you think this is the same issue or something different?

Cheers,

Nick

**Comment by nouiz** [comment]
Open a new issue with the description and the full error message.

Make sure to tell gpu back-end you use and if the other software is in the
same process or not.
